eppo_core/src/events/event_dispatcher.rs

```diff
@@ -0,0 +1,120 @@
+use crate::events::batch_event_processor::BatchEventProcessor;
+use crate::events::event::Event;
+use log::info;
+use std::sync::{Arc, Mutex};
+use tokio::sync::{Notify, OnceCell};
+use tokio::time::{interval, interval_at, Duration, Instant};
+
+#[derive(Debug, Clone)]
+pub struct EventDispatcherConfig {
+    pub sdk_key: String,
+    pub ingestion_url: String,
+    pub delivery_interval_ms: u64,
+    pub retry_interval_ms: u64,
+    pub max_retry_delay_ms: u64,
+    pub max_retries: Option<u32>,
+}
+
+pub struct EventDispatcher {
+    config: EventDispatcherConfig,
+    batch_processor: BatchEventProcessor,
+    delivery_task_active: Arc<Mutex<bool>>,
+}
+
+impl EventDispatcher {
+    pub fn new(config: EventDispatcherConfig, batch_processor: BatchEventProcessor) -> Self {
+        EventDispatcher {
+            config,
+            batch_processor,
+            delivery_task_active: Arc::new(Mutex::new(false)),
+        }
+    }
+
+    /// Enqueues an event in the batch event processor and starts delivery if needed.
+    pub fn dispatch(&self, event: Event) {
+        self.batch_processor.push(event);
+
+        // Start the delivery loop if it's not already active
+        if !self.is_delivery_timer_active() {
+            self.start_delivery_loop();
+        }
```

Comment:
multiple callers could enter here.

consider performing an atomic operation to change from `false` to `true` exactly once. I tried this locally:

<img width="2056" alt="Screenshot 2024-12-20 at 5 17 15 PM" src="https://github.com/user-attachments/assets/43aeda22-068d-4632-a871-ce771e7c8125" />